File: firebase-auth/src/main/java/com/androidhuman/rxfirebase/auth/RxFirebaseUser.java
package com.androidhuman.rxfirebase.auth;

import com.google.firebase.auth.AuthCredential;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.auth.UserProfileChangeRequest;

import android.support.annotation.CheckResult;
import android.support.annotation.NonNull;

import rx.Completable;
import rx.Single;

public final class RxFirebaseUser {

    @CheckResult
    @NonNull
    public static Completable delete(@NonNull FirebaseUser user) {
        return Completable.create(new UserDeleteOnSubscribe(user));
    }

    @CheckResult
    @NonNull
    public static Single<String> getToken(
            @NonNull FirebaseUser user, boolean forceRefresh) {
        return Single.create(new UserGetTokenOnSubscribe(user, forceRefresh));
    }

    @CheckResult
    @NonNull
    public static Single<FirebaseUser> linkWithCredential(
            @NonNull FirebaseUser user, @NonNull AuthCredential credential) {
        return Single.create(new UserLinkWithCredentialOnSubscribe(user, credential));
    }

    @CheckResult
    @NonNull
    public static Completable reauthenticate(
            @NonNull FirebaseUser user, @NonNull AuthCredential credential) {
        return Completable.create(new UserReauthenticateOnSubscribe(user, credential));
    }

    @CheckResult
    @NonNull
    public static Completable reload(@NonNull FirebaseUser user) {
        return Completable.create(new UserReloadOnSubscribe(user));
    }

    @CheckResult
    @NonNull
    public static Completable sendEmailVerification(@NonNull FirebaseUser user) {
        return Completable.create(new UserSendEmailVerificationOnSubscribe(user));
    }

    @CheckResult
    @NonNull
    public static Single<FirebaseUser> unlink(
            @NonNull FirebaseUser user, @NonNull String provider) {
        return Single.create(new UserUnlinkOnSubscribe(user, provider));
    }

    @CheckResult
    @NonNull
    public static Completable updateEmail(
            @NonNull FirebaseUser user, @NonNull String email) {
        return Completable.create(new UserUpdateEmailOnSubscribe(user, email));
    }

    @CheckResult
    @NonNull
    public static Completable updatePassword(
            @NonNull FirebaseUser user, @NonNull String password) {
        return Completable.create(new UserUpdatePasswordOnSubscribe(user, password));
    }

    @CheckResult
    @NonNull
    public static Completable updateProfile(
            @NonNull FirebaseUser user, @NonNull UserProfileChangeRequest request) {
        return Completable.create(new UserUpdateProfileOnSubscribe(user, request));
    }

    private RxFirebaseUser() {
        throw new AssertionError("No instances");
    }
}

File: firebase-auth-kotlin/src/main/kotlin/com/androidhuman/rxfirebase/auth/RxFirebaseUser.kt
@file:Suppress("NOTHING_TO_INLINE")

package com.androidhuman.rxfirebase.auth

import com.google.firebase.auth.AuthCredential
import com.google.firebase.auth.AuthResult
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.auth.UserProfileChangeRequest
import io.reactivex.Completable
import io.reactivex.Single

inline fun FirebaseUser.rxDelete()
        : Completable
        = RxFirebaseUser.delete(this)

inline fun FirebaseUser.rxGetToken(forceRefresh: Boolean)
        : Single<String>
        = RxFirebaseUser.getToken(this, forceRefresh)

inline fun FirebaseUser.rxLinkWithCredential(credential: AuthCredential)
        : Single<AuthResult>
        = RxFirebaseUser.linkWithCredential(this, credential)

inline fun FirebaseUser.rxReauthenticate(credential: AuthCredential)
        : Completable
        = RxFirebaseUser.reauthenticate(this, credential)

inline fun FirebaseUser.rxReload()
        : Completable
        = RxFirebaseUser.reload(this)

inline fun FirebaseUser.rxSendEmailVerification()
        : Completable
        = RxFirebaseUser.sendEmailVerification(this)

inline fun FirebaseUser.rxUnlink(provider: String)
        : Single<AuthResult>
        = RxFirebaseUser.unlink(this, provider)

inline fun FirebaseUser.rxUpdateEmail(email: String)
        : Completable
        = RxFirebaseUser.updateEmail(this, email)

inline fun FirebaseUser.rxUpdatePassword(password: String)
        : Completable
        = RxFirebaseUser.updatePassword(this, password)

inline fun FirebaseUser.rxUpdateProfile(request: UserProfileChangeRequest)
        : Completable
        = RxFirebaseUser.updateProfile(this, request)

